### PR TITLE
feat: linter aspect factories allow customizing rule kinds

### PR DESCRIPTION
**Pull request by alexeagle**
This lets users adapt for situations where they have a custom rule, for example.

### CODE REVIEW

**Comment by alexeagle** [comment]
Note, I didn't touch `clang_tidy.bzl` in this PR because it uses the built-in `CcInfo` provider to determine what to visit.

File: lint/eslint.bzl [hunk 1]
```diff
@@ -247,6 +248,9 @@ def lint_eslint_aspect(binary, configs):
                 allow_single_file = True,
                 cfg = "exec",
             ),
+            "_rule_kinds": attr.string_list(
```
Comment (sallustfire) [inline lint/eslint.bzl:251]: @alexeagle The bazel documentation indicates that implicit attrs must be labels. Is that documentation out of date or in error?

https://bazel.build/rules/lib/globals/bzl.html#aspect

**Comment by jbedard** [comment]
Are "kinds" purely based on the string name of a target? It doesn't matter where it loaded from, macro vs rule etc?

**Comment by alexeagle** [comment]
@jbedard I believe the kind is what appears after the analysis phase, so macros have been expanded already and the `kind` is always what you would see in `bazel query`, which is to say it's the actual rule() declaration.

But yes, if it's named `ts_library` then this matches, whether it's the one from rules_ts or something the user defined.
We could decide to load provider symbols here instead, and then require that the rule provides something. For example ruff and flake8 could be based on a PyInfo provider. But of course we want a provider of sources, and adding a dependency just for that `load` is probably not desirable

File: lint/eslint.bzl (hunk 1 above)
Comment (alexeagle) [reply lint/eslint.bzl:251]: Interesting, I don't recall reading that doc. Apparently it's incorrect - I verified this has the desired behavior.

The real bug is there's no tech writer assigned to the Bazel project. I've been working on fixing it 😉

Comment (sallustfire) [reply lint/eslint.bzl:251]: Good thing you tackled it then. I was thinking that restriction preventing passing the string array via context was going to make the implementation messier than I would have liked.

**Review by jbedard — APPROVED** [review]

File: lint/eslint.bzl (hunk 1 above)
Comment (alexeagle) [reply lint/eslint.bzl:251]: I always start from wishful thinking and then backoff in the face of errors :)

**Pull request merged**